several fixes to scala.concurrent.util.Duration

rkuhn opened the pull request:
- add test cases (migrated from Akka sources)
- add overflow checking (will throw IllegalArgumentException instead of
  giving wrong results)
- make string parsing more precise when giving >100days in nanoseconds
- make method signatures more precise in retaining FiniteDuration
  throughout calculations
- fix mul/div of infinities by negative number
- add Ordering for Deadline (was accidentally left out earlier)

rkuhn committed:
several fixes to scala.concurrent.util.Duration

- add test cases (migrated from Akka sources)
- add overflow checking (will throw IllegalArgumentException instead of
  giving wrong results)
- make string parsing more precise when giving >100days in nanoseconds
- make method signatures more precise in retaining FiniteDuration
  throughout calculations
- fix mul/div of infinities by negative number
- add Ordering for Deadline (was accidentally left out earlier)

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/361/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1069/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1069/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/361/

jsuereth commented:
Reviewers?

retronym commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -289,12 +321,19 @@ class FiniteDuration(val length: Long, val unit: TimeUnit) extends Duration {
+    val c = a + b
+    // check if the signs of the top bit of both summands differ from the sum
+    if (((a ^ c) & (b ^ c)) < 0) throw new IllegalArgumentException("integer overflow")
+    else c
```
I find the technique proposed here a little clearer:

https://www.securecoding.cert.org/confluence/display/java/NUM00-J.+Detect+or+prevent+integer+overflow

rkuhn replied:
It took me about the same time to think through than the one above, but I can change it if you want. None of these are really dead-obvious ;-)

retronym replied:
I suggest to change it and add a comment that links to that CERT page.

rkuhn replied:
okay, will do

rkuhn committed:
second round of Duration cleanup

- make Duration behave consistent with Double's non-finite semantics
- add ScalaDoc
- add complete test suite
- change overflow protection impl after review comments
- clean up code

rkuhn commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/379/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1086/

viktorklang commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -48,9 +129,12 @@ object Duration {
+          case Some(unit) =>
+            val valueStr = s1 dropRight unitName.length
+            val valueD = JDouble.parseDouble(valueStr)
+            if (valueD <= 9007199254740992d && valueD >= -9007199254740992d) Duration(valueD, unit)
```
Magic number is magical. Pull out to a named constant and comment what it is for

rkuhn replied:
will do

jsuereth replied:
I hope you called it MAGIC_NUMBER, for clarity, and this test becomes:  `if(MAGIC_NUMBER is magic)`

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1086/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/379/

viktorklang commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -80,124 +164,217 @@ object Duration {
   val MinusInf: Infinite = new Infinite {
     override def toString = "Duration.MinusInf"
-    def compare(other: Duration) = if (other eq this) 0 else -1
+    def compare(other: Duration) = other match {
```
Is this an improvement over old impl?

rkuhn replied:
no, there was one more case in the meantime and I didn't revert when that went away, will fix

viktorklang commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -80,124 +164,217 @@ object Duration {
+   * Parse String into Duration.  Format is `"<length><unit>"`, where
+   * whitespace is allowed before, between and after the parts. Infinities are
+   * designated by `"Inf"`, `"PlusInf"`, `"+Inf"` and `"-Inf"` or `"MinusInf"`.
+   * Throws exception if format is not parseable.
```
What kind of exception?

rkuhn replied:
good point

viktorklang commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -289,20 +626,56 @@ class FiniteDuration(val length: Long, val unit: TimeUnit) extends Duration {
+    else if (factor.isNaN) Undefined
+    else Zero
+
+  // if this is made a constant, then scalac will elide the conditional and always return +0.0, SI-6331
```
@retronym said he'd take a stab at SI-6331 (for teh record)

rkuhn replied:
which is why I put this comment with the reference in, so that the code can be “normalized” as soon as that issue is fixed

Blaisorblade commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -80,124 +164,217 @@ object Duration {
+      else if (factor < 0d) -this
+      else this
+    def /(factor: Double): Duration =
+      if (factor == 0d || factor.isNaN || factor.isInfinite) Undefined
```
I'd suggest Inf / 0 => Inf instead of Inf/0 => Undefined, consistently with the behavior of /(other: Duration), and being careful with the sign. I'd try to reuse the tricky code for deciding the sign, that is, `(if ((this > Zero) ^ (other >= Zero)) -1 else 1)`.
UPDATE: On second thought, assuming `-0d < 0d`, it's enough to just drop the first case, `factor == 0d ||`.

rkuhn replied:
good point, you found a missing test case; your UPDATE is not enough, though, due to the `-0d` sickness it needs to use `compare` instead of `<`.

BTW: is Double (and Float?) the only example of a violation of the equals/compareTo contract on the JVM which actually ships as part of the platform?

Blaisorblade replied:
Ouch, thanks for the attention. It's good I made my assumption explicit, and bad I couldn't check myself. I agree it's odd that `Double` (and `Float`, yes) give such results, although clearly documented, but AFAIK the equals contract doesn't require consistent results before and after boxing. To me, it's mostly odd that `-0d == 0d` in Java. I'm confused about Scala, where that same expression can ultimately invoke (depending on boxing) `-0d.equals(0d)`.

rkuhn committed:
fix some one more issue in Duration

- Inf / Zero == Inf
- add some more missing test cases
- clarify magic constant
- move exception descriptions into proper @throws docs

Blaisorblade commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -270,9 +611,30 @@ object FiniteDuration {
+      case SECONDS      ⇒ length <= 9223372036L && length >= -9223372036L
+      case MINUTES      ⇒ length <= 153722867L && length >= -153722867L
+      case HOURS        ⇒ length <= 2562047L && length >= -2562047L
+      case DAYS         ⇒ length <= 106751L && length >= -106751L
```
Could these magic constants be made non-magic as well?

rkuhn replied:
Given that the expressed limit is clearly stated in the class’ scaladoc comment, I don’t think it is worth calculating them at some other place and giving them names (which would also slow down the creation of every single Duration instance). And they wouldn’t really get any less magic by moving them into the companion to give them names, because then you’d have to search for their definition instead seeing a value which many programmers will intuitively associate with the right thing (granted, I also don’t know the digits by heart for all of them, but they are easily verified).

So, long story short, the best I can think of is adding comments saying how they were calculated.

Blaisorblade commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -10,36 +10,121 @@ package scala.concurrent.util
+   * '''''Note that on some systems this operation is costly because it entails a system call.'''''
+   * Check `System.nanoTime` for your platform.
+   */
   def isOverdue(): Boolean = (time.toNanos - System.nanoTime()) < 0
```
Why isn't this just timeLeft < Duration.Zero, which is  the specification? Writing it this way introduces some potential for bugs, now or in the future.
The only difference I see is a few object allocations less. Are there platforms where `nanoTime` is fast enough for this to be relevant (making this a non-system-call has been a goal of optimizations in Windows and Linux, I don't remember how successful)?
Of course the result should be the same, but given the possibility of overflow errors being present or introduced by future changes, there should be (outside the ScalaDoc) a comment which explains why this is equivalent, together with (stress?)-tests checking it. After a quick look, I'd propose:
"This is equivalent to timeLeft < Duration.Zero. Since toNanos is guaranteed to never overflow, and Deadline.time is guaranteed to be a positive Duration, there is no risk of overflow and the results are equivalent."
(Note: I didn't check whether there's any overflow risk for `timeLeft < Duration.Zero`, but in the end that code must work since it just uses the public interface as intended).

viktorklang replied:
Because timeLeft allocates a new object.

Blaisorblade replied:
Still, the invariant should be documented and tested. And on second thought, there's a bug already:

```
scala> Duration.Inf.fromNow
res0: scala.concurrent.util.Deadline = Deadline(Duration.Inf)

scala> Duration.Inf.fromNow.isOverdue
java.lang.IllegalArgumentException: toNanos not allowed on infinite Durations
    at scala.concurrent.util.Duration$Infinite$class.toNanos(Duration.scala:171)
    at scala.concurrent.util.Duration$$anon$2.toNanos(Duration.scala:186)
    at scala.concurrent.util.Deadline.isOverdue(Duration.scala:22)
(...)
```

rkuhn replied:
you are right, will fix by requiring FiniteDuration (nothing else makes sense)

daggerrz commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -80,124 +168,218 @@ object Duration {
+      fromNanos((nanos + 0.5).toLong)
+  }
 
+  private val  µs_per_ns = 1000L
```
Making these private[this] would improve the performance of fromNanos. This is s.c.u, after all. :)

paulp replied:
Actually what will really improve it is making them final, which will cause the constants to be inlined directly, at which point the access level doesn't matter since the field will never be accessed.

viktorklang replied:
What's the reason final isn't implied when private?

paulp replied:
Consider a file like this:

```
object Foo {
  private val x = 5
  final val y = 10

  def f1 = () => x
  def f2 = () => y
}
```

f1 returns a closure which returns the value of x.  f2 returns a closure which literally returns "10".  If you recompile this with different values, the closure which returns 10 will still return 10.  The one which accesses x will return the new value of x.

rkuhn replied:
thanks for noticing, making them `private final val` has exactly the right effect (i.e. inline the constants and hide them from Java).

rkuhn committed:
fix two minor issues in Duration

- make constants for fromNanos(Long) actually inlined constants
- clarify origin of require() check constants in FiniteDuration

rkuhn commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/385/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1092/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/385/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1092/

rkuhn commented:
@jsuereth This does not look like my code was even checked out before it failed?

jsuereth commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/387/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1094/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1094/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/387/

viktorklang commented:
@paulp Thanks for the explanation Paul! Isn't that only an issue when you do hot code reloading though?

paulp commented:
It's an issue anytime there is separate compilation.  If someone compiles against library version 1.0 and you ship 1.1, code built against 1.0 will use the final constant type vals from 1.0, not those from 1.1.  This may be desirable or undesirable in any given case, but the point is that it's not the same.

viktorklang commented:
@paulp In the example of the method returning a function returning the constant, that would only be true if the call to the method was replaced with the closure, or?

```
class Foo {
  private final val x = 3
  private val y = 5
  def fX = () => x
  def fY = () => y
}

val f = new Foo
f.fX.apply // call to fX, replacing Foo.class would make this return what the new Foo.class.fX returns?
f.fY.apply
```

paulp commented:
Maybe this will clarify things:

```
// a.scala
class A {
  private val x0 = 1823905
  def x1 = x0
  final val x2 = 1572398
}
// alt-a.scala
class A {
  private val x0 = 0
  def x1 = x0
  final val x2 = 0
}
// b.scala
class B {
  val a = new A
  val y1 = a.x1
  val y2 = a.x2
  override def toString = "(" + y1 + ", " + y2 + ")"
}

object Test {
  def main(args: Array[String]): Unit = {
    println(new B)
  }
}

// demo.sh
#!/bin/sh
#

scalac a.scala b.scala
scala Test
scalac alt-a.scala
scala Test
```

And:

```
% ./demo.sh 
(1823905, 1572398)
(0, 1572398)
```

viktorklang commented:
@paulp Yes, that I fully understand. However, the "private final val"-case works as expected (make x0 private final).

rkuhn commented:
@viktorklang @jsuereth would you please review after my latest changes? Thanks.

jsuereth commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -10,36 +10,121 @@ package scala.concurrent.util
+  def apply(length: Long, unit: String): FiniteDuration   = new FiniteDuration(length,  Duration.timeUnit(unit))
+
+  // Double stores 52 bits mantissa, but there is an implied '1' in front, making the limit 2^53
+  final val maxPreciseDouble = 9007199254740992d 
```
Type ascription police!

Since this is a public api, let's see:

`final val maxPreciseDouble: Double = 9007199254740992d`

Just in case someone ever drops the 'd' from the end in a crazy fit of refactoring.

rkuhn replied:
nope, this needs to be a constant, and unfortunately constants are not allowed to use the police or any armed forces :-(

paulp replied:
Yes, I realize I just said "always specify the return type in public api" but there was an implied exception here, because there is no choice.  Anyway, if someone dropped the d it wouldn't even compile.

jsuereth commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -80,124 +168,218 @@ object Duration {
     override def toString = "Duration.Inf"
-    def compare(other: Duration) = if (other eq this) 0 else 1
+    def compare(other: Duration) = other match {
+      case x if x eq Undefined => -1
```
case `Undefined` => -1  ??

rkuhn replied:
no, Undefined != Undefined (because it is supposed to behave like Double.NaN)

jsuereth commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -80,124 +168,218 @@ object Duration {
-    def compare(other: Duration) = if (other eq this) 0 else 1
+    def compare(other: Duration) = other match {
+      case x if x eq Undefined => -1
+      case x if x eq this      => 0
```
Note: In Fun Hilarity, you can probably safely do:

```
case `Inf` => 0
```

rkuhn replied:
actually it turns out that my version produces a lot less byte code, since your proposal will include code handling the case `Inf == null` (as nonsensical as that may seem)

jsuereth commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -80,124 +168,218 @@ object Duration {
+   * but itself. This value closely corresponds to Double.PositiveInfinity, 
+   * matching its semantics in arithmetic operations.
    */
   val Inf: Infinite = new Infinite {
```
No final?

rkuhn replied:
I’m not aware of any effect this would have; am I wrong?

jsuereth commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -270,9 +611,29 @@ object FiniteDuration {
+       * enforce the 2^63-1 ns limit, must be pos/neg symmetrical because of unary_-
+       */
+      case NANOSECONDS  ⇒ length != Long.MinValue                                     // max_ns = Long.MaxValue
+      case MICROSECONDS ⇒ length <= 9223372036854775L && length >= -9223372036854775L // max_µs = max_ns / 1000
```
Holy magic numbers, batman!   Is there no object we can hide these on as final val?

rkuhn replied:
wouldn’t help, as discussed earlier: they will stay the same numbers, meaning a comment is the best that can be done, and then I’d prefer to have them in-line so people don’t need to go chasing them

jsuereth replied:
Ah, I missed that.   Sounds ok then..., maybe a // about "Don't touch these unless you are (a) God (b) Roland Kuhn (c) have 3 or more reviewers"

paulp replied:
I don't see how a comment is the best which can be done.

```
// a few lines up, it's not much of a chase
final val maxNanos   = Long.MaxValue
final val maxMicros  = maxNanos / 1000
final val maxMillis  = maxMicros / 1000
final val maxSeconds = maxMillis / 1000
final val maxMinutes = maxSeconds / 60
final val maxHours   = maxMinutes / 60
final val maxDays    = maxHours / 24

import FiniteDuration._

require(unit match {
    /*
     * enforce the 2^63-1 ns limit, must be pos/neg symmetrical because of unary_-
     */
    case NANOSECONDS  ⇒   -maxNanos <= length && length <= maxNanos
    case MICROSECONDS ⇒  -maxMicros <= length && length <= maxMicros
    case MILLISECONDS ⇒  -maxMillis <= length && length <= maxMillis
    case SECONDS      ⇒ -maxSeconds <= length && length <= maxSeconds
    case MINUTES      ⇒ -maxMinutes <= length && length <= maxMinutes
    case HOURS        ⇒   -maxHours <= length && length <= maxHours
    case DAYS         ⇒    -maxDays <= length && length <= maxDays
    case _ ⇒
      val v = DAYS.convert(length, unit)
      -maxDays <= v && v <= maxDays
  }, "Duration is limited to +-(2^63-1)ns (ca. 292 years)")
```

You don't have to hardcode long strings of digits.

rkuhn replied:
Indeed, it turns out that my worries about the constants turning public were unfounded, and the overhead of one `lneg` in each case surely gets lost in the noise. I’m running the tests again and will push a fix shortly.

jsuereth commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -289,20 +650,56 @@ class FiniteDuration(val length: Long, val unit: TimeUnit) extends Duration {
+
+  // see https://www.securecoding.cert.org/confluence/display/java/NUM00-J.+Detect+or+prevent+integer+overflow
+  private def safeAdd(a: Long, b: Long): Long = {
+    if ((b > 0) && (a > Long.MaxValue - b) ||
```
```
`<joke>` Can we put ZERO in a constant to avoid this magic number? `</joke>`
```

jsuereth commented on test/files/jvm/duration-tck.scala:
```diff
@@ -0,0 +1,191 @@
+
+  implicit class Assert(val left: Any) extends AnyVal {
+    import Duration.Undefined
+    def mustBe(right: Any) = right match {
```
Sneaking some scalatest in?

rkuhn replied:
well, the test was ported from Akka and I didn’t want to change it more than necessary

jsuereth commented on test/files/jvm/duration-tck.scala:
```diff
@@ -0,0 +1,191 @@
+    }
+  }
+
+  def intercept[T <: Exception : ClassTag](code: => Unit) =
```
This is a highly useful method.  Note: You can define abstract traits in src/partest that you can extend in these tests.   I'd love to see intercept there.

Also note:  `scala.util.control.Exception.catching` should probably look like this a bit too.

rkuhn replied:
yes, I did that in https://github.com/scala/scala/pull/1279, but since these branches are not based upon each other I cannot use it from here.

jsuereth replied:
Gotcha.  We'll have to do an intercept refactor later.

Blaisorblade replied:
Notice: I remember at least another use of intercept (and friends) in FutureTests, TryTests, etc.

jsuereth commented:
These changes are a great improvement!  I have a few minor comments, then LGTM.

rkuhn committed:
two more minor cleanups to Duration

- copy partest TestUtil.intercept change from PR 1279 branch
- add comment on non-obvious match cases

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/412/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1118/

rkuhn committed:
factor out more constants in Duration, thanks paulp!

rkuhn commented:
PLS REBUILD ALL

rkuhn committed:
retronym had one more improvement hidden up his sleeve for Duration

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1122/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/416/

rkuhn committed:
make private methods private[this] to avoid conflicts when importing Duration._

Blaisorblade commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -10,36 +10,121 @@ package scala.concurrent.util
+ * does not take into account changes to the system clock (such as leap
+ * seconds).
+ */
+case class Deadline private (time: Duration) extends Ordered[Deadline] {
```
It seems that `time` is in fact required to be a `FiniteDuration`, because e.g. `isOverdue` fails if `time` is not a `FiniteDuration`; if so, this should be checked statically by making `time`'s type `FiniteDuration`. Consequently, you'd have to move `fromNow` into `FiniteDuration` to make it typecheck.
If instead `Inf.fromNow` is supposed to be legal, then `isOverdue` is buggy.

rkuhn replied:
Thanks, I did exactly as you proposed (move fromNow into FiniteDuration) and added a negative test so that `(x: Duration).fromNow` does not compile.

rkuhn committed:
restrict Deadline to finite durations (would have to throw otherwise)

rkuhn commented:
PLS REBUILD ALL

rkuhn commented:
@jsuereth since you merged #1279 yesterday, there will be a (logical) conflict now between this and master. Shall I merge master in here and fix it up?

rkuhn committed:
Merge remote-tracking branch 'origin/2.10.x' into fix-duration-issues-RK

rkuhn committed:
improve Promise.tryAwait by converting to match statement

- also use type FiniteDuration due to a previous change to
  Deadline’s type signature

rkuhn commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/419/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1126/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1126/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/419/

jsuereth commented:
Thanks for the hard work and cleanup!
